File: 1/2020.09.14/155.Min Stack/python/solution.py
class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.data = []
        self.min = []

    def push(self, x: int) -> None:
        self.data.append(x)
        if not self.min or x <= self.min[-1]:
            self.min.append(x)

    def pop(self) -> None:
        x = self.data.pop()
        if x == self.min[-1]:
            self.min.pop()

    def top(self) -> int:
        return self.data[-1] 

    def getMin(self) -> int:
        return self.min[-1]

# only use one stack
class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.data = []
        self.MIN = None


    def push(self, x: int) -> None:
        if self.MIN is None or x <= self.MIN:
            self.data.append(self.MIN)
            self.MIN = x
        self.data.append(x)


    def pop(self) -> None:
        r = self.data.pop()
        if r == self.MIN:
            self.MIN = self.data.pop()


    def top(self) -> int:
        return self.data[-1]


    def getMin(self) -> int:
        return self.MIN

File: 1/2020.09.14/155.Min Stack/java/solution.java
class MinStack {

	private Node head;
    /** initialize your data structure here. */
    public MinStack() {

    }
    
    public void push(int x) {
		if (head == null) {
			head = new Node(x, x);
		} else {
			head = new Node(x, Math.min(x, head.min), head);
		}
    }
    
    public void pop() {
       head = head.next; 
    }
    
    public int top() {
       return head.val; 
    }
    
    public int getMin() {
       return head.min; 
    }

	class Node {
		int val;
		int min;
		Node next;
		public Node(int val, int min) {
			this(val, min, null);
		}

		public Node(int val, int min, Node next) {
			this.val = val;
			this.min = min;
			this.next = next;
		}
	}
}
